### ISSUE TITLE
Weird colors in the output of my attack

### ISSUE THREAD

**Issue by SamHSlva** [Harry24k/adversarial-attacks-pytorch#11]
I am using PyTorch CIFAR-10 dataset, with the following transformations:

transform_train = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
    ])

The clean images are showing like this:
<img width="347" alt="Screen Shot 2020-11-09 at 5 26 46 PM" src="https://user-images.githubusercontent.com/36551435/98608457-d6e74000-22b0-11eb-94e9-e40ee34b966c.png">
The attacked images with attack = CW(net, c=0.0004, kappa=0, steps=10, lr=0.01), are like this:
<img width="347" alt="Screen Shot 2020-11-09 at 5 27 06 PM" src="https://user-images.githubusercontent.com/36551435/98608518-0007d080-22b1-11eb-9c20-39034bd85f17.png">

From the original paper, we expect the images to have few to none visual difference, but I am seeing a big color offset. Do you have any hint on what I could be doing wrong in my code?

Thank you for the library, it is of GREAT HELP.

**Comment by Harry24k** [comment]
All images should be scaled to [0, 1] with transform[to.Tensor()] before used in attacks. To make it easy to use adversarial attacks, a reverse-normalization is not included in the attack process. To apply an input normalization, please add a normalization layer to the model. Please refer to the [demo](https://nbviewer.jupyter.org/github/Harry24k/adversarial-attacks-pytorch/blob/master/demos/White%20Box%20Attack%20%28Imagenet%29.ipynb).

**Comment by SamHSlva** [comment]
Thank you a lot for your reply. 
I've normalized the way you've suggested, with the normalization layer added to my model as:
class Normalize(nn.Module) :
```
    def __init__(self, mean, std) :
        super(Normalize, self).__init__()
        self.register_buffer('mean', torch.Tensor(mean))
        self.register_buffer('std', torch.Tensor(std))


   def forward(self, input):
        # Broadcasting
        mean = self.mean.reshape(1, 3, 1, 1)
        std = self.std.reshape(1, 3, 1, 1)
        return (input - mean) / std


norm_layer = Normalize(mean=[0.4914, 0.4822, 0.4465], std=[0.2023, 0.1994, 0.2010])
transform_test = transforms.Compose([transforms.ToTensor(),])
testset = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=transform_test)
testloader = torch.utils.data.DataLoader(
    testset, batch_size=32, shuffle=False, num_workers=2)
net = WideResNet(depth=34, num_classes=10, widen_factor=10, dropRate=0.0)
net = net.to(device)
if device == 'cuda':
    net = torch.nn.DataParallel(net)
    cudnn.benchmark = True
checkpoint = torch.load('ckpt.pth')
net.load_state_dict(checkpoint)
model = nn.Sequential(norm_layer, net).to(device)
```

With the attack still as: `attack = CW(net, c=0.0004, kappa=0, steps=10, lr=0.01)`

My clean images still show:
<img width="233" alt="Screen Shot 2020-11-09 at 11 00 57 PM" src="https://user-images.githubusercontent.com/36551435/98629874-791e1c80-22df-11eb-84fd-be7f8234e417.png">
And my attacked images show:
<img width="236" alt="Screen Shot 2020-11-09 at 11 00 49 PM" src="https://user-images.githubusercontent.com/36551435/98629892-83d8b180-22df-11eb-8ef5-29353236a7eb.png">

Am I still doing something wrong?

Thank you very much for your time, and, again for making this library available.

**Comment by Harry24k** [comment]
Please try `attack = CW(model, c=0.0004, kappa=0, steps=10, lr=0.01)` instead of `attack = CW(net, c=0.0004, kappa=0, steps=10, lr=0.01)`

**Comment by SamHSlva** [comment]
I've done as suggested with no luck. I've tried FGSM also, which is giving me some more realistic images.

Clean output:

<img width="544" alt="Screen Shot 2020-11-10 at 9 51 09 AM" src="https://user-images.githubusercontent.com/36551435/98697512-a21fcb00-233a-11eb-8140-5561392eecac.png">

FGSM output: `attack = FGSM(model, eps=0.04)`

<img width="547" alt="Screen Shot 2020-11-10 at 9 50 39 AM" src="https://user-images.githubusercontent.com/36551435/98697308-66850100-233a-11eb-8acf-191aa026f9f3.png">

CW output: attack = `CW(model, c=0.04, kappa=0, steps=10, lr=0.01)`

<img width="546" alt="Screen Shot 2020-11-10 at 9 50 58 AM" src="https://user-images.githubusercontent.com/36551435/98697544-a9df6f80-233a-11eb-867d-3afd59a3bb0b.png">




```
import torch
import torch.backends.cudnn as cudnn
import torchvision
import torchvision.transforms as transforms
from torch.utils.data import DataLoader, TensorDataset, Dataset
from models import *
from models.madry_model import WideResNet
import my_torchattacks.attack as torchattacks
from my_torchattacks import PGD, FGSM, CW, BIM
from scheduler import CycleScheduler
from os import listdir
from scipy.spatial.distance import mahalanobis
import torchvision.utils
from torchvision import models
import torchvision.datasets as dsets
import torchvision.transforms as transforms
import os
import sys

import matplotlib.pyplot as plt
%matplotlib inline

os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"   # see issue #152
os.environ["CUDA_VISIBLE_DEVICES"]="1"
class Normalize(nn.Module) :
    def __init__(self, mean, std) :
        super(Normalize, self).__init__()
        self.register_buffer('mean', torch.Tensor(mean))
        self.register_buffer('std', torch.Tensor(std))
        
    def forward(self, input):
        # Broadcasting
        mean = self.mean.reshape(1, 3, 1, 1)
        std = self.std.reshape(1, 3, 1, 1)
        return (input - mean) / std
norm_layer = Normalize(mean=[0.4914, 0.4822, 0.4465], std=[0.2023, 0.1994, 0.2010])

device = 'cuda' if torch.cuda.is_available() else 'cpu'
transform_train = transforms.Compose([transforms.ToTensor(),])


testset = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=transform_train)
testloader = torch.utils.data.DataLoader(
    testset, batch_size=32, shuffle=False, num_workers=2)

model = VGG('VGG16')
model = model.to(device)
if device == 'cuda':
    model = torch.nn.DataParallel(model)
    cudnn.benchmark = True
checkpoint = torch.load('checkpoint/VGG16_original_randomnoise_0.3.pth')
model.load_state_dict(checkpoint)
model = nn.Sequential(norm_layer, model).to(device)
model.eval()

criterion = nn.CrossEntropyLoss()
attack = CW(model, c=0.04, kappa=0, steps=10, lr=0.01)
# attack = FGSM(model, eps=0.04)
test_loss = 0
total = 0
correct=0
for batch_idx, (inputs, targets) in enumerate(testloader):
    inputs, targets = inputs, targets.to(device)
    adv_images = attack(inputs.to(device), targets)
    inputs2 = inputs
    inputs = adv_images
    outputs, _ = model(inputs)
    loss = criterion(outputs, targets)
    test_loss += loss.item()
    _, predicted = outputs.max(1)
    total += targets.size(0)
    print("\r{} completed: {:.2%}".format(
            'Testing', batch_idx / len(testloader)), end="")
    sys.stdout.flush()
    accuracy = 100.*correct/total
    break
    if (batch_idx / len(testloader))>=0.01:
        break
# print('Gamma', gamma)
print('Accuracy', 100.*correct/total)

grid_img = torchvision.utils.make_grid(inputs.cpu(), nrow=8, normalize=True, range=(inputs.cpu().min(),inputs.cpu().max()))
plt.figure(figsize=(10,5))
plt.imshow(grid_img.permute(1, 2, 0))

grid_img = torchvision.utils.make_grid(inputs2.cpu(), nrow=8, normalize=True, range=(inputs2.cpu().min(),inputs2.cpu().max()))
plt.figure(figsize=(10,5))
plt.imshow(grid_img.permute(1, 2, 0))
```

My model:

Sequential(
  (0): Normalize()
  (1): DataParallel(
    (module): VGG(
      (features): Sequential(
        (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): ReLU(inplace=True)
        (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (4): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (5): ReLU(inplace=True)
        (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        (7): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (8): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (9): ReLU(inplace=True)
        (10): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (11): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (12): ReLU(inplace=True)
        (13): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        (14): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (15): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (16): ReLU(inplace=True)
        (17): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (18): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (19): ReLU(inplace=True)
        (20): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (21): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (22): ReLU(inplace=True)
        (23): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        (24): Conv2d(256, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (25): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (26): ReLU(inplace=True)
        (27): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (28): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (29): ReLU(inplace=True)
        (30): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (31): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (32): ReLU(inplace=True)
        (33): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        (34): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (35): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (36): ReLU(inplace=True)
        (37): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (38): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (39): ReLU(inplace=True)
        (40): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        (41): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (42): ReLU(inplace=True)
        (43): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        (44): AvgPool2d(kernel_size=1, stride=1, padding=0)
      )
      (classifier): Linear(in_features=512, out_features=10, bias=True)
    )
  )
)

**Comment by Harry24k** [comment]
I guess the steps of CW is too low.

Considering that the initial image is `w = torch.zeros_like(images)`, we need to set enough steps to make a similar image to the original.

Here are the results of different steps:
`torchattacks.CW(model, c=0.01, kappa=0, steps=10, lr=0.01),`
![1](https://user-images.githubusercontent.com/41545927/98831642-3004ce80-247f-11eb-83a9-fdc63a91b497.png)
`torchattacks.CW(model, c=0.01, kappa=0, steps=100, lr=0.01),`
![2](https://user-images.githubusercontent.com/41545927/98831668-35faaf80-247f-11eb-93ed-47432357452d.png)
`torchattacks.CW(model, c=0.01, kappa=0, steps=1000, lr=0.01),`
![3](https://user-images.githubusercontent.com/41545927/98831679-38f5a000-247f-11eb-86ae-32e2d6b63028.png)

**Comment by SamHSlva** [comment]
That makes sense. 
Thank you for your help. I've increased the number of steps and got way better images.
